src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/MemberExpression.cs

```diff
@@ -302,6 +303,8 @@ public static MemberExpression Property(Expression? expression, MethodInfo prope
             return Property(expression, GetProperty(propertyAccessor, nameof(propertyAccessor)));
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
+            Justification = "Since the propertyAccessor is being supplied, the Property won't be trimmed.")]
```

Comment:
This suppression is super sketchy. Given a MethodInfo accessor, find the PropertyInfo for it. While that in itself might work, if someone tries grabbing the setter for the PropertyInfo, it might have been trimmed.

Opinions on this? Should I mark the usages of this method as "unsafe"? Or is this OK to suppress?

cc @vitek-karas @MichalStrehovsky

Reply:
I think this should be unsafe. We've been discussing removing property metadata - which would break this for example.

Currently the logic is: The only way to get `PropertyInfo` is through reflection APIs which are either unsafe, or handled by linker (or through annotations). If it's handled, then linker guarantees presence of the property metadata and both getter and setter.
But it doesn't work the other way round, having MethodInfo on a getter doesn't guarantee anything about the setter  - and I don't think it should guarantee anything about the property either.

Reply:
The problem is, `GetProperty` is called from the method directly above it:

https://github.com/dotnet/runtime/blob/e6453ecf2ef5ca92d52f8b3bfa364bc832c3a4e8/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/MemberExpression.cs#L292-L303

And that is the method that is called from C# when you do something like the following:

```C#
        IQueryable<Customer> customers = GetCustomers();
        IEnumerable<Customer> invalidCustomers = customers.Where(c => string.IsNullOrEmpty(c.Name));
```

From the IL of the above code:

```
	IL_0000: call class [System.Linq.Expressions]System.Linq.IQueryable`1<class Customer> Program::GetCustomers()
	IL_0005: ldtoken Customer
	IL_000a: call class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
	IL_000f: ldstr "c"
	IL_0014: call class [System.Linq.Expressions]System.Linq.Expressions.ParameterExpression [System.Linq.Expressions]System.Linq.Expressions.Expression::Parameter(class [System.Runtime]System.Type, string)
	IL_0019: stloc.0
	IL_001a: ldnull
	IL_001b: ldtoken method bool [System.Runtime]System.String::IsNullOrEmpty(string)
	IL_0020: call class [System.Runtime]System.Reflection.MethodBase [System.Runtime]System.Reflection.MethodBase::GetMethodFromHandle(valuetype [System.Runtime]System.RuntimeMethodHandle)
	IL_0025: castclass [System.Runtime]System.Reflection.MethodInfo
	IL_002a: ldc.i4.1
	IL_002b: newarr [System.Linq.Expressions]System.Linq.Expressions.Expression
	IL_0030: dup
	IL_0031: ldc.i4.0
	IL_0032: ldloc.0
	𝐈𝐋_𝟎𝟎𝟑𝟑: 𝐥𝐝𝐭𝐨𝐤𝐞𝐧 𝐦𝐞𝐭𝐡𝐨𝐝 𝐢𝐧𝐬𝐭𝐚𝐧𝐜𝐞 𝐬𝐭𝐫𝐢𝐧𝐠 𝐂𝐮𝐬𝐭𝐨𝐦𝐞𝐫::𝐠𝐞𝐭_𝐍𝐚𝐦𝐞()
	𝐈𝐋_𝟎𝟎𝟑𝟖: 𝐜𝐚𝐥𝐥 𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐞𝐟𝐥𝐞𝐜𝐭𝐢𝐨𝐧.𝐌𝐞𝐭𝐡𝐨𝐝𝐁𝐚𝐬𝐞 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐞𝐟𝐥𝐞𝐜𝐭𝐢𝐨𝐧.𝐌𝐞𝐭𝐡𝐨𝐝𝐁𝐚𝐬𝐞::𝐆𝐞𝐭𝐌𝐞𝐭𝐡𝐨𝐝𝐅𝐫𝐨𝐦𝐇𝐚𝐧𝐝𝐥𝐞(𝐯𝐚𝐥𝐮𝐞𝐭𝐲𝐩𝐞 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞𝐌𝐞𝐭𝐡𝐨𝐝𝐇𝐚𝐧𝐝𝐥𝐞)
	𝐈𝐋_𝟎𝟎𝟑𝐝: 𝐜𝐚𝐬𝐭𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐞𝐟𝐥𝐞𝐜𝐭𝐢𝐨𝐧.𝐌𝐞𝐭𝐡𝐨𝐝𝐈𝐧𝐟𝐨
	𝐈𝐋_𝟎𝟎𝟒𝟐: 𝐜𝐚𝐥𝐥 𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬]𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬.𝐌𝐞𝐦𝐛𝐞𝐫𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬]𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧::𝐏𝐫𝐨𝐩𝐞𝐫𝐭𝐲(𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬]𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧, 𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐞𝐟𝐥𝐞𝐜𝐭𝐢𝐨𝐧.𝐌𝐞𝐭𝐡𝐨𝐝𝐈𝐧𝐟𝐨)
	IL_0047: stelem.ref
	IL_0048: call class [System.Linq.Expressions]System.Linq.Expressions.MethodCallExpression [System.Linq.Expressions]System.Linq.Expressions.Expression::Call(class [System.Linq.Expressions]System.Linq.Expressions.Expression, class [System.Runtime]System.Reflection.MethodInfo, class [System.Linq.Expressions]System.Linq.Expressions.Expression[])
	IL_004d: ldc.i4.1
	IL_004e: newarr [System.Linq.Expressions]System.Linq.Expressions.ParameterExpression
	IL_0053: dup
	IL_0054: ldc.i4.0
	IL_0055: ldloc.0
	IL_0056: stelem.ref
	IL_0057: call class [System.Linq.Expressions]System.Linq.Expressions.Expression`1<!!0> [System.Linq.Expressions]System.Linq.Expressions.Expression::Lambda<class [System.Runtime]System.Func`2<class Customer, bool>>(class [System.Linq.Expressions]System.Linq.Expressions.Expression, class [System.Linq.Expressions]System.Linq.Expressions.ParameterExpression[])
	IL_005c: call class [System.Linq.Expressions]System.Linq.IQueryable`1<!!0> [System.Linq.Queryable]System.Linq.Queryable::Where<class Customer>(class [System.Linq.Expressions]System.Linq.IQueryable`1<!!0>, class [System.Linq.Expressions]System.Linq.Expressions.Expression`1<class [System.Runtime]System.Func`2<!!0, bool>>)
	IL_0061: pop
	IL_0062: ret
```

So if I marked this method as unsafe, I would need to mark `Property(Expression? expression, MethodInfo propertyAccessor)` as unsafe. Which means anyone writing the above code would get an ILLink warning.

Maybe we might need to special case `Expression::Property(Expression, MethodInfo propertyAccessor)` in the mono/linker and make it preserve the whole property?

Reply:
sigh... I give up... I probably hate the least to make this a special case in the linker. We already have special cases for some of the `Expression` stuff: https://github.com/mono/linker/blob/319ed02ed9a25ec3800f723ff72a1eb92ef050c7/src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs#L241-L244

It's kind of funny: Expression.Property(Type, name) is "Safer" than Expression.Property(Type, getterMethodInfo).

Reply:
Note that it won't be as simple as that - to make this work linker will have to track at least on the basic level `MethodInfo`.

Reply:
I've opened https://github.com/mono/linker/issues/1814 for this and changed the annotation to be `RequiresUnreferencedCode`.